### PR TITLE
avoid circular dependency usage between mcore and modelopt

### PR DESCRIPTION
**Pull request by dimapihtar**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Remerge of #5644 (#6247).

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by AAnoosheh — APPROVED** [review]

**Review by Phlip79 — APPROVED** [review]

**Comment by maanug-nv** [comment]
### Code review

Found 1 issue:

1. Preserve ModelOpt optionality when saving checkpoints (bug; prior pull request context)

The `has_nvidia_modelopt` guards were removed, but both replacement helpers import `modelopt` before checking whether the model was converted. Any ordinary `torch_dist` or legacy checkpoint save from a supported installation without optional `nvidia-modelopt` now aborts with `ModuleNotFoundError` (the distributed path does so after writing the main state). Please retain the guards or make the helpers tolerate a missing dependency.

https://github.com/NVIDIA/Megatron-LM/blob/2e81ccf11cab2296bb8568929704bc1bb5c2d7a1/megatron/training/checkpointing.py#L858-L860

https://github.com/NVIDIA/Megatron-LM/blob/2e81ccf11cab2296bb8568929704bc1bb5c2d7a1/megatron/training/checkpointing.py#L912-L917

https://github.com/NVIDIA/Megatron-LM/blob/2e81ccf11cab2296bb8568929704bc1bb5c2d7a1/megatron/core/post_training/modelopt/checkpointing.py#L47-L58

<sub>Generated by Codex.</sub>

**Review by maanug-nv — APPROVED** [review]

File: megatron/core/post_training/modelopt/checkpointing.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    # Early return if the model already has a modelopt_state or the checkpoint does not exist.
+    if not os.path.exists(modelopt_checkpoint_name) or mto.ModeloptStateManager.is_converted(
+        model[0]
+    ):
+        return
+
+    # Loading the common modelopt_state (replicated on all ranks).
+    # Detect format: legacy checkpoints store common state in a standalone common.pt file;
+    # newer sharded checkpoints store it as a ShardedObject inside the torch_dist checkpoint.
+    legacy_common_path = os.path.join(modelopt_checkpoint_name, "common.py")
```
Comment (cuichenx) [inline megatron/core/post_training/modelopt/checkpointing.py]: typo?

File: megatron/core/post_training/modelopt/checkpointing.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    if os.path.exists(legacy_common_path):
+        common_modelopt_state = safe_load_from_bytes(legacy_common_path)
```
Comment (cuichenx) [inline megatron/core/post_training/modelopt/checkpointing.py]: this is treating the path string as bytes

File: megatron/core/post_training/modelopt/checkpointing.py (hunk 1 above)
Comment (dimapihtar) [reply megatron/core/post_training/modelopt/checkpointing.py]: fixed

File: megatron/core/post_training/modelopt/checkpointing.py (hunk 2 above)
Comment (dimapihtar) [reply megatron/core/post_training/modelopt/checkpointing.py]: fixed, no need in additional check anymore. `load_common_state_dict` handles it.

**Review by cuichenx — APPROVED** [review]
LGTM

**Review by jaredcasper — APPROVED** [review]

File: megatron/core/post_training/modelopt/checkpointing.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    modelopt_state = copy.deepcopy(mto.modelopt_state(model[0]))
+    remove_per_module_state(modelopt_state)
+    save(modelopt_state, modelopt_checkpoint_name, sharded_strategy)
+
+
+def _load_extra_state_from_sharded_checkpoint(
+    model: torch.nn.Module,
+    checkpoint_name: str | Path,
+    prefix: str,
+    metadata: dict[str, Any] | None = None,
```
Comment (cuichenx) [inline megatron/core/post_training/modelopt/checkpointing.py]: metadata is not used?

Comment (dimapihtar) [reply megatron/core/post_training/modelopt/checkpointing.py]: looks like it. Will remove
